### ISSUE TITLE
Dict[A, B] isn't compatible with Dict[A, Union[B, C]]

### ISSUE THREAD

**Issue by WieeRd** [microsoft/pyright#2254]
**To Reproduce**
```python
a: Dict[int, Union[int, str]]
b: Dict[int, str] = {}

a = b
```

**Expected behavior**
`A` is compatible with `Union[A, B]` so I expected it to work as dictionary value as well
```python
a: Union[str, int]
b: str = "Hello World"

a = b # this doesn't produce error
```

**Screenshots or Code**
![image](https://user-images.githubusercontent.com/59733058/131785332-cce2a767-082d-4cc8-8205-05f1ef54baa8.png)

**VS Code extension or command-line**
I'm using pyright inside Neovim, with help of [coc.nvim](https://github.com/neoclide/coc.nvim) plugin and [coc-pyright](https://github.com/fannheyward/coc-pyright) package.

**Comment by erictraut** [comment]
This is correct behavior. The type parameters for `Dict` are invariant, so you cannot assign a `Dict[int, str]` to a `Dict[int, int | str]`. Here's why this isn't allowed. Let's say that you follow your above example with this code.

```python
a[3] = 4
```

Because `b` and `a` refer to the same dictionary, now `b`'s type has been violated because it is supposed to contain only `str` values.

For more details, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#generic-types) or PEP 483 and 484.

**Comment by WieeRd** [comment]
Well it makes sense than. Thanks.
